nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrUtils.java

```diff
@@ -280,5 +291,115 @@ public SolrInputDocument toSolrInputDocument(SolrDocument d) {
+                final String stringValue = DataTypeUtils.toString(coercedValue, () -> DataTypeUtils.getDateFormat(RecordFieldType.DATE.getDefaultFormat()));
+                if (DataTypeUtils.isLongTypeCompatible(stringValue)) {
+                    LocalDate localDate = getLocalDateFromEpochTime(fieldName, coercedValue);
+                    inputDocument.addField(fieldName,localDate.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)+'Z');
+                } else {
+                    inputDocument.addField(fieldName,LocalDate.parse(stringValue).format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)+'Z');
+                }
+                break;
+            }
+            case TIMESTAMP: {
+                final String stringValue = DataTypeUtils.toString(coercedValue, () -> DataTypeUtils.getDateFormat(RecordFieldType.TIMESTAMP.getDefaultFormat()));
+                if (DataTypeUtils.isLongTypeCompatible(stringValue)) {
+                    LocalDateTime localDateTime = getLocalDateTimeFromEpochTime(fieldName, coercedValue);
+                    inputDocument.addField(fieldName,localDateTime.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)+'Z');
+                } else {
+                    inputDocument.addField(fieldName,LocalDateTime.parse(stringValue).format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)+'Z');
+                }
+                break;
+            }
+            case DOUBLE:
+                inputDocument.addField(fieldName,DataTypeUtils.toDouble(coercedValue, fieldName));
+                break;
+            case FLOAT:
+                inputDocument.addField(fieldName,DataTypeUtils.toFloat(coercedValue, fieldName));
+                break;
+            case LONG:
+                inputDocument.addField(fieldName,DataTypeUtils.toLong(coercedValue, fieldName));
+                break;
+            case INT:
+            case BYTE:
+            case SHORT:
+                inputDocument.addField(fieldName,DataTypeUtils.toInteger(coercedValue, fieldName));
+                break;
+            case CHAR:
+            case STRING:
+                inputDocument.addField(fieldName,coercedValue.toString());
+                break;
+            case BIGINT:
+                if (coercedValue instanceof Long) {
+                    inputDocument.addField(fieldName,(Long) coercedValue);
+                } else {
+                    inputDocument.addField(fieldName,(BigInteger) coercedValue);
+                }
+                break;
+            case BOOLEAN:
+                final String stringValue = coercedValue.toString();
+                if ("true".equalsIgnoreCase(stringValue)) {
+                    inputDocument.addField(fieldName,true);
+                } else if ("false".equalsIgnoreCase(stringValue)) {
+                    inputDocument.addField(fieldName,false);
+                } else {
+                    inputDocument.addField(fieldName,stringValue);
+                }
+                break;
+            case RECORD: {

```

Comment:
How do nested records end up being represented in the Solr document? Not saying anything is wrong here, just asking to understand how it works.

Lets say we have a person schema with top-level fields for "firstName" and "lastName", and "address", and the address field is of type record and then has it's own fields "street", "city", "zip"...  

Does the resulting Solr document contain "firstName", "lastName", "street", "city", "zip"? 

Would it make sense to have an option to include the parent field in the field names, so it ends up being "address_street", "address_city", and "address_zip" so that you know where those fields came from?

Reply:
Yes, I've implemented it to flatten the nested record into a single solr document, although i was sort of unsure if there would be a need to consider nested records. If there is one, I'll make changes to the field name as you suggested so that one knows where the field came from.

Reply:
I think we have to handle it since someone can specify a field name in "fields to index" that could be of type record.

I think it makes sense to have a property like "Nested Field Names" with choices for "Fully Qualified" and "Child Only" (or something like that).

This lines up with how Solr's JSON update works:

https://lucene.apache.org/solr/guide/6_6/transforming-and-indexing-custom-json.html#transforming-and-indexing-custom-json

The part that shows....

The default behavior is to use the fully qualified name (FQN) of the node. So, if we don’t define any field mappings, like this:

curl 'http://localhost:8983/solr/my_collection/update/json/docs?split=/exams'\
    -H 'Content-type:application/json' -d '
{
  "first": "John",
  "last": "Doe",
  "grade": 8,
  "exams": [
    {
      "subject": "Maths",
      "test"   : "term1",
      "marks"  : 90},
    {
      "subject": "Biology",
      "test"   : "term1",
      "marks"  : 86}
  ]
}'

The indexed documents would be added to the index with fields that look like this:

{
  "first":"John",
  "last":"Doe",
  "grade":8,
  "exams.subject":"Maths",
  "exams.test":"term1",
  "exams.marks":90},
{
  "first":"John",
  "last":"Doe",
  "grade":8,
  "exams.subject":"Biology",
  "exams.test":"term1",
  "exams.marks":86}

Reply:
This also brings up another scenario... what do we do if there is an array field, and the type of the elements in the array is a record?

That would be similar to the "exams" array in the above example. With Solr's JSON update handler you would have to say split=/exams and this produces a Solr document for each exam. 

I'm actually not sure what Solr does if you left off the split param because then you would have multiple fields with the same name in the same document, like exams.subject would be there twice.

Reply:
Do you want me to replicate the top level field in N nested record of the array and convert into into N solr documents, but this might get complicated if there are more than one levels of nested records.

Reply:
Well with the code you already have, what does the Solr document look like when there is an array of records?

Reply:
With the current code, it writes a single solr document for a record and flattens all the nested records in that single solr document.
So if there is an array of nested records it would create multiple fields with the same key in the solr document which would eventually mean that the field would be indexed as multivalued in solr with the assumption that the schema has defined the field to be multivalued else it would fail to index.

Reply:
I think the current approach is probably fine for now, maybe we can make a note in the @CapabilityDescription that mentions how nested records are handled, or in the additionalDetails.html you could provide example input and show it would be handled.

Someone can always use ConvertRecord to convert whatever format to JSON, and then use the existing PutSolrContentStream with Solr's splitting if they want to do something like that.

Reply:
Ok, I'll then just change the field names as discussed before and I'll leave the implementation as it is and give some examples as part of the additionalDetails.html

Reply:
I've added examples around the behavior of nested records in the additionalDetails.html and I've modified the field names to include the parent field name